File: java/src/main/java/com/zcash/t2z/T2zLib.java
/**
 * JNA interface for the t2z native library.
 *
 * This class provides low-level FFI bindings to the Rust t2z library.
 * Users should prefer the high-level API in T2z.java instead of using these directly.
 */
package com.zcash.t2z;

import com.sun.jna.*;
import com.sun.jna.ptr.PointerByReference;
import java.util.Arrays;
import java.util.List;

/**
 * JNA interface to the t2z native library.
 */
public interface T2zLib extends Library {

    T2zLib INSTANCE = Native.load(
        Platform.isMac() ? "t2z" : Platform.isLinux() ? "t2z" : "t2z",
        T2zLib.class
    );

    // Result codes
    int SUCCESS = 0;
    int ERROR_NULL_POINTER = 1;
    int ERROR_INVALID_UTF8 = 2;
    int ERROR_BUFFER_TOO_SMALL = 3;
    int ERROR_PROPOSAL = 10;
    int ERROR_PROVER = 11;
    int ERROR_VERIFICATION = 12;
    int ERROR_SIGHASH = 13;
    int ERROR_SIGNATURE = 14;
    int ERROR_COMBINE = 15;
    int ERROR_FINALIZATION = 16;
    int ERROR_PARSE = 17;
    int ERROR_NOT_IMPLEMENTED = 99;

    // Error handling
    int pczt_get_last_error(byte[] buffer, NativeLong bufferLen);

    // Transaction request
    int pczt_transaction_request_new(
        CPayment.ByReference payments,
        NativeLong numPayments,
        PointerByReference requestOut
    );

    void pczt_transaction_request_free(Pointer request);

    int pczt_transaction_request_set_target_height(Pointer request, int targetHeight);

    int pczt_transaction_request_set_use_mainnet(Pointer request, boolean useMainnet);

    // PCZT operations
    int pczt_propose_transaction(
        byte[] inputsBytes,
        NativeLong inputsBytesLen,
        Pointer request,
        String changeAddress,
        PointerByReference pcztOut
    );

    int pczt_prove_transaction(Pointer pczt, PointerByReference pcztOut);

    int pczt_verify_before_signing(
        Pointer pczt,
        Pointer request,
        CTransparentOutput.ByReference expectedChange,
        NativeLong expectedChangeLen
    );

    int pczt_get_sighash(Pointer pczt, NativeLong inputIndex, byte[] sighashOut);

    int pczt_append_signature(
        Pointer pczt,
        NativeLong inputIndex,
        byte[] signature,
        PointerByReference pcztOut
    );

    int pczt_combine(Pointer[] pczts, NativeLong numPczts, PointerByReference pcztOut);

    int pczt_finalize_and_extract(
        Pointer pczt,
        PointerByReference txBytesOut,
        NativeLongByReference txBytesLenOut
    );

    int pczt_parse(byte[] pcztBytes, NativeLong pcztBytesLen, PointerByReference pcztOut);

    int pczt_serialize(Pointer pczt, PointerByReference bytesOut, NativeLongByReference bytesLenOut);

    void pczt_free(Pointer pczt);

    void pczt_free_bytes(Pointer bytes, NativeLong len);

    /**
     * C-compatible payment structure for JNA.
     */
    @Structure.FieldOrder({"address", "amount", "memo", "label", "message"})
    class CPayment extends Structure {
        public String address;
        public long amount;
        public String memo;
        public String label;
        public String message;

        public CPayment() {
            super();
        }

        public static class ByReference extends CPayment implements Structure.ByReference {
        }

        public static class ByValue extends CPayment implements Structure.ByValue {
        }
    }

    /**
     * C-compatible transparent output structure for JNA.
     */
    @Structure.FieldOrder({"script_pub_key", "script_pub_key_len", "value"})
    class CTransparentOutput extends Structure {
        public Pointer script_pub_key;
        public NativeLong script_pub_key_len;
        public long value;

        public CTransparentOutput() {
            super();
        }

        public static class ByReference extends CTransparentOutput implements Structure.ByReference {
        }

        public static class ByValue extends CTransparentOutput implements Structure.ByValue {
        }
    }

    /**
     * JNA NativeLong by reference helper.
     */
    class NativeLongByReference extends com.sun.jna.ptr.ByReference {
        public NativeLongByReference() {
            super(NativeLong.SIZE);
        }

        public NativeLong getValue() {
            if (NativeLong.SIZE == 4) {
                return new NativeLong(getPointer().getInt(0));
            } else {
                return new NativeLong(getPointer().getLong(0));
            }
        }

        public void setValue(NativeLong value) {
            if (NativeLong.SIZE == 4) {
                getPointer().setInt(0, value.intValue());
            } else {
                getPointer().setLong(0, value.longValue());
            }
        }
    }
}

File: kotlin/src/main/kotlin/com/zcash/t2z/T2zLib.kt
/**
 * JNA interface for the t2z native library
 *
 * This file provides low-level FFI bindings to the Rust t2z library.
 * Users should prefer the high-level API in T2z.kt instead of using these directly.
 */
package com.zcash.t2z

import com.sun.jna.*
import com.sun.jna.ptr.PointerByReference

/**
 * Result codes from FFI functions
 */
enum class ResultCode(val code: Int) {
    SUCCESS(0),
    ERROR_NULL_POINTER(1),
    ERROR_INVALID_UTF8(2),
    ERROR_BUFFER_TOO_SMALL(3),
    ERROR_PROPOSAL(10),
    ERROR_PROVER(11),
    ERROR_VERIFICATION(12),
    ERROR_SIGHASH(13),
    ERROR_SIGNATURE(14),
    ERROR_COMBINE(15),
    ERROR_FINALIZATION(16),
    ERROR_PARSE(17),
    ERROR_NOT_IMPLEMENTED(99);

    companion object {
        fun fromCode(code: Int): ResultCode =
            entries.find { it.code == code } ?: throw IllegalArgumentException("Unknown result code: $code")
    }
}

/**
 * C-compatible payment structure for JNA
 */
@Structure.FieldOrder("address", "amount", "memo", "label", "message")
open class CPayment : Structure() {
    @JvmField var address: String? = null
    @JvmField var amount: Long = 0
    @JvmField var memo: String? = null
    @JvmField var label: String? = null
    @JvmField var message: String? = null

    class ByReference : CPayment(), Structure.ByReference
    class ByValue : CPayment(), Structure.ByValue
}

/**
 * C-compatible transparent output structure for JNA
 */
@Structure.FieldOrder("script_pub_key", "script_pub_key_len", "value")
open class CTransparentOutput : Structure() {
    @JvmField var script_pub_key: Pointer? = null
    @JvmField var script_pub_key_len: NativeLong = NativeLong(0)
    @JvmField var value: Long = 0

    class ByReference : CTransparentOutput(), Structure.ByReference
    class ByValue : CTransparentOutput(), Structure.ByValue
}

/**
 * JNA interface to the t2z native library
 */
interface T2zLib : Library {
    companion object {
        val INSTANCE: T2zLib by lazy {
            val libName = when {
                Platform.isMac() -> "t2z"
                Platform.isLinux() -> "t2z"
                Platform.isWindows() -> "t2z"
                else -> throw UnsupportedOperationException("Unsupported platform")
            }
            Native.load(libName, T2zLib::class.java)
        }
    }

    // Error handling
    fun pczt_get_last_error(buffer: ByteArray, bufferLen: NativeLong): Int

    // Transaction request
    fun pczt_transaction_request_new(
        payments: Array<CPayment>?,
        numPayments: NativeLong,
        requestOut: PointerByReference
    ): Int

    fun pczt_transaction_request_free(request: Pointer?)

    fun pczt_transaction_request_set_target_height(request: Pointer, targetHeight: Int): Int

    fun pczt_transaction_request_set_use_mainnet(request: Pointer, useMainnet: Boolean): Int

    // PCZT operations
    fun pczt_propose_transaction(
        inputsBytes: ByteArray,
        inputsBytesLen: NativeLong,
        request: Pointer,
        changeAddress: String?,
        pcztOut: PointerByReference
    ): Int

    fun pczt_prove_transaction(pczt: Pointer, pcztOut: PointerByReference): Int

    fun pczt_verify_before_signing(
        pczt: Pointer,
        request: Pointer,
        expectedChange: Array<CTransparentOutput>?,
        expectedChangeLen: NativeLong
    ): Int

    fun pczt_get_sighash(pczt: Pointer, inputIndex: NativeLong, sighashOut: ByteArray): Int

    fun pczt_append_signature(
        pczt: Pointer,
        inputIndex: NativeLong,
        signature: ByteArray,
        pcztOut: PointerByReference
    ): Int

    fun pczt_combine(pczts: Array<Pointer>, numPczts: NativeLong, pcztOut: PointerByReference): Int

    fun pczt_finalize_and_extract(
        pczt: Pointer,
        txBytesOut: PointerByReference,
        txBytesLenOut: NativeLongByReference
    ): Int

    fun pczt_parse(pcztBytes: ByteArray, pcztBytesLen: NativeLong, pcztOut: PointerByReference): Int

    fun pczt_serialize(pczt: Pointer, bytesOut: PointerByReference, bytesLenOut: NativeLongByReference): Int

    fun pczt_free(pczt: Pointer?)

    fun pczt_free_bytes(bytes: Pointer?, len: NativeLong)
}

/**
 * JNA NativeLong by reference helper
 */
class NativeLongByReference : com.sun.jna.ptr.ByReference(NativeLong.SIZE) {
    fun getValue(): NativeLong = NativeLong(when (NativeLong.SIZE) {
        4 -> getPointer().getInt(0).toLong()
        8 -> getPointer().getLong(0)
        else -> throw IllegalStateException("Unexpected NativeLong.SIZE: ${NativeLong.SIZE}")
    })

    fun setValue(value: NativeLong) {
        when (NativeLong.SIZE) {
            4 -> getPointer().setInt(0, value.toInt())
            8 -> getPointer().setLong(0, value.toLong())
            else -> throw IllegalStateException("Unexpected NativeLong.SIZE: ${NativeLong.SIZE}")
        }
    }
}
